Option to kill other resque-pool instances on startup

joshuaflanagan opened the pull request:
Adds the `--kill-others` command line option.
When this option is set, resque-pool will send a graceful
shutdown signal to all other running resque-pools.

This is useful in "no downtime deploy" scenarios,
where you want the pool running the new code to
completely load its environment and be ready to
process work *before* shutting down the old code.
Once the new code is ready, it will shut down the
old processes for you.

See also https://github.com/nevans/resque-pool/pull/132

brasic commented:
`--kill-others` implies `--no-pidfile`, since a second daemon could never get to `kill_other_pools!` without it.  It might make sense to make the dependency explicit by either failing fast or making `--kill-others` set `--no-pidfile` if it is not already.

joshuaflanagan commented:
That is a good point. Although, I think you *could* run multiple instances with pidfiles by specifying different pidfile paths. Probably not a common scenario, but maybe enough to not force the `--not-pidfile` option. Might be enough to mention it in the README that you probably want to use the options together.

nevans commented:
I very much like the idea of defaulting `--no-pidfile --lock tmp/resque-pool.lock` when `--kill-others` is set, especially since you can subsequently override either (e.g. `--kill-others --pidfile /var/resque-pool.pid --lock /var/resque-pool.lock`).

nevans commented on lib/resque/pool/killer.rb:
```diff
@@ -0,0 +1,41 @@
+
+
+      def all_resque_pool_processes
+        out, err, status = Open3.capture3("ps ax")
```
This ought to always work just fine on Linux/Mac OS X/BSD...  but my paranoia prefers to use `ps -e -o pid= -o command=` (POSIX flags, restricted output columns).  Your current `parse_pids_from_output` should work just fine on either, I think.

nevans commented on lib/resque/pool/killer.rb:
```diff
@@ -0,0 +1,41 @@
+      end
+
+      def parse_pids_from_output(output)
+        pool_lines = output.split("\n").grep(/resque-pool-master/)
```
Or you could make a stricter parser that first splits each line into two, and only matches `resque-pool-master` at the beginning of the first column.  This will avoid false positives like `grep resque-pool-master /var/log/app.log`.  Extremely rare in practice, but highly surprising when it bites you.

joshuaflanagan commented:
@nevans I've addressed your concerns regarding the `ps` parsing. We'll need to get https://github.com/nevans/resque-pool/pull/132 merged before I add the logic around `--no-pidfile`.

joshuaflanagan commented:
What if we add an additional flag that encompasses the many options that work together for this approach? For example, we could add --min-downtime (or --max-uptime), which sets --kill-others, --no-pidfile and --lock. That might be more discoverable than having --kill-others imply all of those options. I like the idea of having discrete flags that do one thing, but also a flag to group the flags commonly used together.

nevans commented:
I'm okay with that approach.  I think I prefer `--hot-swap` (or something similar) to `--min-downtime`.  What do you think?
